core/services/headtracker/head_listener.go

```diff
@@ -0,0 +1,258 @@
+				hl.logger().Error("HeadTracker: got nil block header")
+				continue
+			}
+			promNumHeadsReceived.Inc()
+
+			err := handleNewHead(ctx, *blockHeader)
+			if ctx.Err() != nil {
+				// the 'ctx' context is closed only on ht.done - on shutdown, so it's safe to return nil
+				return nil
+			} else if err != nil {
+				return err
+			}
+		case err, open := <-hl.headSubscription.Err():
+			if open && err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// subscribe periodically attempts to connect to the ethereum node via websocket.
+// It returns true on success, and false if cut short by a done request and did not connect.
+func (hl *HeadListener) subscribe(connected func()) bool {
+	hl.sleeper.Reset()
+	for {
+		if err := hl.unsubscribeFromHead(); err != nil {
+			hl.logger().Error("failed when unsubscribe from head", err)
+			return false
+		}
+
+		hl.logger().Info("HeadListener: Connecting to ethereum node ", hl.config.EthereumURL(), " in ", hl.sleeper.Duration())
+		select {
+		case <-hl.chStop:
+			return false
+		case <-time.After(hl.sleeper.After()):
+			err := hl.subscribeToHead(connected)
+			if err != nil {
+				promEthConnectionErrors.Inc()
+				hl.logger().Warnw(fmt.Sprintf("HeadListener: Failed to connect to ethereum node %v", hl.config.EthereumURL()), "err", err)
+			} else {
+				hl.logger().Info("HeadListener: Connected to ethereum node ", hl.config.EthereumURL())
+				return true
+			}
+		}
+	}
+}
+
+func (hl *HeadListener) subscribeToHead(connected func()) error {
+	hl.connectedMutex.Lock()
+	defer hl.connectedMutex.Unlock()
+
+	hl.headers = make(chan *models.Head)
+
+	sub, err := hl.ethClient.SubscribeNewHead(context.Background(), hl.headers)
+	if err != nil {
+		return errors.Wrap(err, "EthClient#SubscribeNewHead")
+	}
+	if err := verifyEthereumChainID(hl); err != nil {
+		return errors.Wrap(err, "verifyEthereumChainID failed")
+	}
+
+	hl.headSubscription = sub
+	hl.connected = true
+
+	connected()
+	return nil
+}
+
+func (hl *HeadListener) unsubscribeFromHead() error {
+	hl.connectedMutex.Lock()
+	defer hl.connectedMutex.Unlock()
+
+	if !hl.connected {
+		return nil
+	}
+
+	timedUnsubscribe(hl.headSubscription)
+	hl.connected = false
+
+	// ht.headers will be nil if subscription failed, channel closed, and
+	// receiveHeaders returned from the loop. listenForNewHeads will set it to
+	// nil in that case to avoid a double close panic.
+	if hl.headers != nil {
+		close(hl.headers)
+	}
+	return nil
+}
+
+// timedUnsubscribe attempts to unsubscribe but aborts abruptly after a time delay
+// unblocking the application. This is an effort to mitigate the occasional
+// indefinite block described here from go-ethereum:
+// https://chainlink/pull/600#issuecomment-426320971
+func timedUnsubscribe(unsubscriber ethereum.Subscription) {
+	unsubscribed := make(chan struct{})
+	go func() {
+		unsubscriber.Unsubscribe()
+		close(unsubscribed)
+	}()
+	select {
+	case <-unsubscribed:
+	case <-time.After(100 * time.Millisecond):
```

Comment:
I reckon we could increase this

Reply:
500 ms? did you see any timeouts around it on certain chains?

Reply:
Oops, I was thinking of a different part of the code. Pls ignore

Reply:
👍🏼